### ISSUE TITLE
Failed to create resource. You can't create this secret because a secret with this name is already scheduled for deletion

### ISSUE THREAD

**Issue by mariotoffia** [udondan/cdk-ec2-key-pair#2]
Hi, I've been updating my CDK project to include your excellent CDK library. However, I've stumbeled into a problem. I did create a keypair in secrets manager and downloaded it sucessfully. I've undeployed and tried to re-deploy and got that the secret was **marked for deletion**. I cannot see it in the secrets manager console. Is it a must to generate a unque name each time in order to not get into this conflict?

Cheers,
 Mario :)

## CDK output:
```
EC2KeyPaireedefaultkeypairMario0599CC7A) Resource creation Initiated
 41/61 | 8:41:08 AM | CREATE_FAILED        | Custom::EC2-Key-Pair                           | ee-default-keypair-Mario/EC2-Key-Pair-ee-default-keypair-Mario/Default (eedefaultkeypairMario
EC2KeyPaireedefaultkeypairMario0599CC7A) Failed to create resource. You can't create this secret because a secret with this name is already scheduled for deletion. | Full error in CloudWa
tch 2020/04/03/[$LATEST]3a3efa6d32f545bf82c323db4100bf5d
        new CustomResource (/c/progs/aws/greengrass-starter/node_modules/@aws-cdk/aws-cloudformation/lib/custom-resource.ts:163:21)
        \_ new KeyPair (/c/progs/aws/greengrass-starter/node_modules/cdk-ec2-key-pair/lib/index.ts:93:21)
        \_ new EdgeAnalyticsStack (/c/progs/aws/greengrass-starter/lib/edge-analytics-stack.ts:594:20)
        \_ Object.<anonymous> (/c/progs/aws/greengrass-starter/bin/edge-analytics.ts:29:1)
        \_ Module._compile (internal/modules/cjs/loader.js:1139:30)
        \_ Module.m._compile (/c/progs/aws/greengrass-starter/node_modules/ts-node/src/index.ts:837:23)
        \_ Module._extensions..js (internal/modules/cjs/loader.js:1159:10)
        \_ Object.require.extensions.<computed> [as .ts] (/c/progs/aws/greengrass-starter/node_modules/ts-node/src/index.ts:840:12)
        \_ Module.load (internal/modules/cjs/loader.js:988:32)
        \_ Function.Module._load (internal/modules/cjs/loader.js:896:14)
        \_ Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)
        \_ main (/c/progs/aws/greengrass-starter/node_modules/ts-node/src/bin.ts:226:14)
        \_ Object.<anonymous> (/c/progs/aws/greengrass-starter/node_modules/ts-node/src/bin.ts:485:3)
        \_ Module._compile (internal/modules/cjs/loader.js:1139:30)
        \_ Object.Module._extensions..js (internal/modules/cjs/loader.js:1159:10)
        \_ Module.load (internal/modules/cjs/loader.js:988:32)
        \_ Function.Module._load (internal/modules/cjs/loader.js:896:14)
        \_ Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)
        \_ /usr/local/lib/node_modules/npm/node_modules/libnpx/index.js:268:14
```

**Comment by udondan** [comment]
Hi Mario,

yes, that's the case for secrets in the secrets manager. You cannot directly delete a secret. You can schedule deletion with a minimum period of 7 days. Until it finally got deleted you cannot create a new secret with the same name.

In the console you can see secrets marked for deletion by clicking the little engine wheel and then mark the checkbox in the settings pane:

![Settings](https://user-images.githubusercontent.com/6443408/78334067-5f6b6880-758b-11ea-9e0d-6a7117a98634.png)

Then, when opening the secret, you're even able to cancel the scheduled deletion.

My construct uses the minimum period of 7 days to schedule deletion. I have it on my roadmap to support automatic name generation with a random postfix, as most of the core constructs do, to avoid the need to manually change the name.

Originally I had the idea to also support SSM param store as a backend, but dropped that. Maybe I should revisit that. SSM params can also be stored 

Cheers,
Daniel

**Comment by mariotoffia** [comment]
Thanks for the prompt response - I'm postfixing with an uuid and uses that instead.

Cheers,
 Mario :)

**Comment by mariotoffia** [comment]
Just a question - instead of just creating a secret - if that fail couldn't you update that for example in [custom-certificate-lambda.py - line 60](https://github.com/aws-samples/aws-iot-greengrass-edge-analytics-workshop/blob/master/lib/custom-certificate-handler/custom-certificate-lambda.py#L60) they try to create a secret - but if that fails they try to update it instead (when exception is `ResourceExistsException`).

snippet:
```py
            try:
                response = secretsmanager.create_secret(
                    Name=secretName, SecretString=json.dumps(credentials))
            except ClientError as e:
                if e.response['Error']['Code'] == 'ResourceExistsException':
                    response = secretsmanager.update_secret(
                        SecretId=secretName, SecretString=json.dumps(credentials))
```

and when they delete, they use `ForceDeleteWithoutRecovery` set to `True`. Is that possible for you too?
[line 82](https://github.com/aws-samples/aws-iot-greengrass-edge-analytics-workshop/blob/master/lib/custom-certificate-handler/custom-certificate-lambda.py#L82):
```py
            secretsmanager.delete_secret(
                SecretId=secretName, ForceDeleteWithoutRecovery=True)
```

**Comment by udondan** [comment]
Ouh, I missed that option! Thanks! I think the `ForceDeleteWithoutRecovery` will be enough. It shouldn't update the secret if it already exists. You then could accidentally override the key from a different stack.

**Comment by mariotoffia** [comment]
Maybe an option `allowOverwrite` that default to `false`? In my case, I actually want to update... :)

**Comment by udondan** [comment]
I just released version [v1.1.0](https://github.com/udondan/cdk-ec2-key-pair/releases/tag/v1.1.0).

The `ForceDeleteWithoutRecovery` option now is implemented and also it is the default. It probably will be a rare case where someone wants to actually keep the secret longer than the Key Pair itself.

> Maybe an option `allowOverwrite` that default to false? In my case, I actually want to update... :)

Can you explain the use case? It seems quite strange to me. This would enable you to manage the same resource from multiple stacks. From what you had described above the now implemented functionality should be enough. You now can delete the resource and re-create it with the same name.
